pkg/base-dev/lib/noun-diff.hoon

```diff
@@ -0,0 +1,227 @@
+      %atom  empty-set
+    ==
+  ++  find-ins-holes
+    |=  diff=ins-diff
+    ~+
+    ^-  (set @)
+    ?-  -.diff
+      %hole  (silt ~[+<.diff])
+      %cell
+        (~(uni in $(diff lhs.diff)) $(diff rhs.diff))
+      %atom  empty-set
+    ==
+  ++  filter-del-holes
+    |=  [allowed-holes=(set @) diff=del-diff]
+    ^-  del-diff
+    ?:  ?=(%ignore -.diff)  diff
+    ?-  -.diff
+      %hole
+        ?:  (~(has in allowed-holes) +.diff)  diff
+        [%ignore ~]
+      %cell  [%cell $(diff +<.diff) $(diff +>.diff)]
+    ==
+  ++  filter-ins-holes
+    |=  [allowed-holes=(set @) diff=ins-diff]
+    ^-  final-ins-diff
+    ?-  -.diff
+      %hole
+        ?:  (~(has in allowed-holes) +<.diff)  [%hole +<.diff]
+        (insify-noun original:diff)
+      %cell  [%cell $(diff +<.diff) $(diff +>.diff)]
+      %atom  diff
+    ==
+  ++  gcp
+    |=  diff=diff
+    ^-  patch
+    ?-  -.ins.diff
+      %atom  
+        [%diff diff]
+      %cell
+        ?:  ?=(%cell -.del.diff)
+        [%cell $(diff [+<.del.diff +<.ins.diff]) $(diff [+>.del.diff +>.ins.diff])]
+        [%diff diff]
+      %hole  
+        [%diff diff]
+    ==
+  ++  closure
+    |=  =patch
+    ^-  [^patch ?]
+    ?-  -.patch
+      %diff
+        =/  del-holes  (find-del-holes del:patch)
+        =/  ins-holes  (find-final-ins-holes ins:patch)
+        =/  difference  (~(dif in ins-holes) del-holes)
+        [patch =(difference empty-set)]
+      %cell
+        =/  lhs  $(patch lhs:patch)
+        =/  rhs  $(patch rhs:patch)
+        ?:  ?&(+.lhs +.rhs)  [[%cell -.lhs -.rhs] &]
+        $(patch (pull-diff [%cell -.lhs -.rhs]))
+    ==
+  ++  pull-diff
+    |=  =patch
+    ^-  [%diff del=del-diff ins=final-ins-diff]
+    ?-  -.patch
+      %diff  patch
+      %cell
+        =/  pulled-lhs  $(patch lhs:patch)
+        =/  pulled-rhs  $(patch rhs:patch)
+        :+
+          %diff 
+          [%cell del:pulled-lhs del:pulled-rhs]
+          [%cell ins:pulled-lhs ins:pulled-rhs]
+    ==
+  ++  is-subtree  
+    |=  [tree=* subtree=*]
+    ~+
+    ^-  ?
+    ?:  =(tree subtree)  &
+    ?@  tree  |
+    ?|
+      (is-subtree -.tree subtree)
+      (is-subtree +.tree subtree)
+    ==
+  ++  oracle
+    |=  [a=* b=*]
+    |=  subtree=*
+    ^-  (unit @)
+    ?: 
+    ?&
+      (is-subtree a subtree)
+      (is-subtree b subtree)
+    ==
+    `(mug subtree)
+    ~
+  ++  extract-del
+    |=  [oracle=$-(* (unit @)) subtree=*]
+    ~+
+    ^-  del-diff
+    =/  hash  (oracle subtree)
+    ?^  hash  [%hole +.hash]
+    ?@  subtree  [%ignore ~]
+    [%cell (extract-del oracle -.subtree) (extract-del oracle +.subtree)]
+  ++  extract-ins
+    |=  [oracle=$-(* (unit @)) subtree=*]
+    ~+
+    ^-  ins-diff
+    =/  hash  (oracle subtree)
+    ?^  hash  [%hole +.hash subtree]
+    ?@  subtree  [%atom subtree]
+    [%cell (extract-ins oracle -.subtree) (extract-ins oracle +.subtree)]
+  ++  ins 
+    |=  [diff=final-ins-diff var-map=(map @ *)]
+    ^-  *
+    ?-  -.diff
```

Comment:
These are the two standard styles for `?-`s:

For long lines, use this:
```hoon
?-    -.diff
    %atom
  +.diff
::
    %cell
  [$(diff +<.diff) $(diff +>.diff)]
::
    ::  ...
==
```
For shorter lines, use this:
```hoon
?-  -.diff
  %atom  +.diff
  %cell  [$(diff +<.diff) $(diff +>.diff)]
  %hole  (~(got by var-map) +.diff)
==
```